05-html-css/flexbox-intro.md

```diff
@@ -0,0 +1,230 @@
+
+![Cross Axis: row](./imgs/flexbox-axis-cross-row.png)
+
+![Cross Axis: column](./imgs/flexbox-axis-cross-column.png)
+
+You can determine the direction of the main axis with the property `flex-direction`.
+
+`flex-direction` has the following properties:
+- `row` (default)
+- `row-reverse`
+- `column`
+- `column-reverse`
+
+**Practice:** Pair up with someone around you and make a prediction for what each of these will do. Then, add the property `flex-direction` to [the Codepen](https://codepen.io/adadev/pen/jzNQmV?editors=1100) and change the values and observe the differences.
+
+## Determining Wrapping Behavior
+
+We can see in our example right now that our flex items extend beyond the flex container. Let's change that: we can determine the behavior for what our flex items should do if it fills the flex container along the main axis using `flex-wrap` on the flex container:
+
+```css
+.container {
+  display: flex;
+  flex-wrap: wrap;
+}
+```
+
+The available values for `flex-wrap` are:
+- `nowrap`
+- `wrap`
+- `wrap-reverse`
+
+**Practice:** Using [the same Codepen](https://codepen.io/adadev/pen/jzNQmV?editors=1100), add the property `flex-wrap` and each of the values to observe the differences.
+
+## Controlling Alignment
+
+Let's look at a modification of the example we've been following. [Look at this Codepen](https://codepen.io/adadev/pen/ZxzVLq?editors=1100); it's nearly identical, but we've added some stuff that essentially shows a more realistic case where every flex item is not identical in width or height. This will help us see empty space between flex items better.
+
+![Varied Width and Height Items, unaligned](./imgs/flexbox-alignment-example-1.png)
+
+How are the flex items aligned against each other right now? Where is there empty space/negative space? Why is it there? What kind of alignment would be more pleasant to look at/be a better experience for a website?
+
+To control alignment of flex items within the flex container, we have two main properties: `justify-content` and `align-items`.
+
+`justify-content` determines alignment along the main axis. Its possible values are:
+- `stretch`
+- `flex-start`
+- `flex-end`
+- `center`
+- `space-around`
+- `space-between`
+- `space-evenly`
+
+`align-items` determines alignment along the cross axis. Its possible values are:
+- `stretch`
+- `flex-start`
+- `flex-end`
+- `center`
+
+Take the time to see each value of both properties in action.
+
+**Optional:** What properties work best to achieve the following look?
+
+![Varied Width and Height Items, aligned](./imgs/flexbox-alignment-example-2.png)
+
+
+### Experiment: Check for Responsive Design
+
+We'll spend time talking about responsive design later. For now, let's spend some time asking ourselves the question: "does this _work_ with different browser sizes?" By _work_, we can ask "does this look good?" but we must always ask "is the site still functional?"
+
```

Comment:
Maybe don't lead with "we'll cover this later". This section might be stronger as:

----------

**Responsive design** is all about answering the question, "does this _work_ with different browser sizes?" By _work_, we can ask "does this look good?" but we must always ask "is the site still functional?"

Using flexbox happens to be very compatible with thinking and practicing responsive design on web design.

For every flexbox example we've looked at today, change the `.container`'s width to `50vw` instead of `500px`, then resizes the browser and observe how the flex items behave. _Why does that work?_

Because of the rapid rise of mobile browsing in recent years, responsive design is an increasingly important part of modern web development. We will return to it in more depth in a future lesson.